File: generator/generator.py
import time
import math
from math import sqrt, floor
import pandas as pd
import pyarrow as pa
import pyarrow.parquet as pq

# 1 Milliarde
GOAL = 1e8

# 168

precomputed_primes = []

# Simple Sieve of Erat.
def simpleSieve(limit):
    mark = [True] * (limit + 1)
    p = 2

    while (p * p <= limit):
        if (mark[p] == True): 
            for i in range(p * p, limit + 1, p): 
                mark[i] = False  

        p += 1
        
    for p in range(2, limit + 1): 
        if mark[p]:
            precomputed_primes.append(p)


def get_chunks(limit):
    ranges = []
    # Chunk size
    size = int(GOAL / 100)
    start = 2

    while start < limit:
        end = int(min(start + size, limit))
        ranges.append((start, end))
        start += size
    
    return ranges

def segmentedSieve(min, max):
    primes = [True] * (max - min + 1)

    for precomp in precomputed_primes:
        first_multiple = (min // precomp)

        if first_multiple <= 1:
            first_multiple = precomp + precomp
        elif (min % precomp) != 0:
            first_multiple = (first_multiple * precomp) + precomp
        else:
            first_multiple = first_multiple * precomp

        for j in range(first_multiple, max + 1, precomp):
            primes[j - min] = False
            
    final_primes = []
    for k in range(min, max+ 1):
            if primes[k - min]:
                final_primes.append(k)

    return final_primes

def is_prime(prime):
    if prime in precomputed_primes:
        return True

    for precomp in precomputed_primes:
        if prime % precomp == 0:
            return False
    return True


def get_twin_prime(prime, prime_list):
    """Check if prime is part of a twin prime pair and return the twin if it exists"""
    if prime - 2 in precomputed_primes or prime - 2 in prime_list:
        return prime - 2
    elif prime + 2 in prime_list:
        return prime + 2
    return None


def check_safe_prime(prime, prime_list):
    """Check if (p-1)/2 is also prime"""
    safe_candidate = (prime - 1) // 2
    return safe_candidate in precomputed_primes or safe_candidate in prime_list


def get_mersenne_k(prime):
    """Check if prime is a Mersenne prime candidate (2^k - 1) and return k if true"""
    # A more efficient check: if prime = 2^k - 1, then prime + 1 = 2^k
    # So prime + 1 must be a power of 2, which means (prime + 1) & prime == 0
    if (prime + 1) & prime == 0:
        return int(math.log2(prime + 1))
    return None


# Process each prime to its other formats and properties
def map_prime_properties(prime_list, start_index=0):
    """
    Maps each prime number to its properties as specified in the requirements.
    
    Args:
        prime_list: List of prime numbers to process
        start_index: Starting index for the prime numbers in the sequence
        
    Returns:
        DataFrame with all the requested properties
    """
    # Initialize lists for each property
    data = {
        'index': [],
        'prime': [],
        'gap_to_previous': [],
        'twin_prime': [],
        'is_safe_prime': [],
        'mersenne_k': [],
        'bit_length': [],
        'digit_sum_base10': [],
        'digit_sum_base2': [],
        'is_palindrome_base10': [],
        'div_mod_10': []
    }
    
    # Set the previous prime for gap calculation
    prev_prime = 2 if start_index == 0 else precomputed_primes[start_index - 1]
    
    for i, prime in enumerate(prime_list):
        # index - position in the sequence
        index = start_index + i
        data['index'].append(index)
        
        # prime - the prime itself
        data['prime'].append(prime)
        
        # gap_to_previous - difference from the previous prime
        gap = prime - prev_prime
        data['gap_to_previous'].append(gap)
        prev_prime = prime
        
        # twin_prime - twin prime if one exists
        twin = get_twin_prime(prime, prime_list)
        data['twin_prime'].append(twin)
        
        # is_safe_prime - Is (p-1)/2 also prime?
        is_safe = check_safe_prime(prime, prime_list)
        data['is_safe_prime'].append(is_safe)
        
        # mersenne_k - If Mersenne candidate: which k? (else NULL)
        mersenne_k = get_mersenne_k(prime)
        data['mersenne_k'].append(mersenne_k)
        
        # bit_length - Number of bits needed to represent prime
        bit_length = prime.bit_length()
        data['bit_length'].append(bit_length)
        
        # digit_sum_base10 - Sum of digits in base 10
        digit_sum_base10 = sum(int(digit) for digit in str(prime))
        data['digit_sum_base10'].append(digit_sum_base10)
        
        # digit_sum_base2 - Sum of digits in base 2
        digit_sum_base2 = bin(prime).count('1')
        data['digit_sum_base2'].append(digit_sum_base2)
        
        # is_palindrome_base10 - Is the prime a palindrome?
        prime_str = str(prime)
        is_palindrome = prime_str == prime_str[::-1]
        data['is_palindrome_base10'].append(is_palindrome)
        
        # div_mod_10 - For exploratory analysis (e.g., does it end with 1, 3, 7, 9?)
        div_mod_10 = prime % 10
        data['div_mod_10'].append(div_mod_10)
    
    return data

# Define the schema for our parquet file
def create_parquet_schema():
    return pa.schema([
        ('index', pa.int32()),
        ('prime', pa.int64()),
        ('gap_to_previous', pa.int32()),
        ('twin_prime', pa.int32()),
        ('is_safe_prime', pa.bool_()),
        ('mersenne_k', pa.int32()),
        ('bit_length', pa.int32()),
        ('digit_sum_base10', pa.int32()),
        ('digit_sum_base2', pa.int32()),
        ('is_palindrome_base10', pa.bool_()),
        ('div_mod_10', pa.int32())
    ])

# Initialize the parquet writer
def initialize_parquet_writer(output_file):
    schema = create_parquet_schema()
    return pq.ParquetWriter(output_file, schema)

# Write a batch of data to the parquet file
def write_batch_to_parquet(writer, data):
    # Convert the dictionary to a PyArrow Table
    table = pa.Table.from_pydict(data, schema=writer.schema)
    writer.write_table(table)

start = time.time()

print(f"Starting data generation with a goal of {int(GOAL):_}...")

up_to_sqrt = floor(sqrt(GOAL)) + 1

print(f"Precomputing primes up to square root of goal: {up_to_sqrt:_}...")

# Compute up to sqrt(limit) with simple sieve
simpleSieve(up_to_sqrt)

print(f"Split goal into chunks of ranges...")

# Chunk the limit
chunks = get_chunks(GOAL)

print(f"{len(chunks)} chunks of ranges computed...")

print(chunks)

# Initialize the parquet writer
output_file = "primes_data.parquet"
parquet_writer = initialize_parquet_writer(output_file)

total = 0
current_index = 0

for chunk_index, chunk in enumerate(chunks):
    print(f"Processing chunk {chunk_index+1}/{len(chunks)}: {chunk}...")

    # Generate primes for this chunk
    primes = segmentedSieve(*chunk)
    chunk_size = len(primes)
    total += chunk_size
    
    # Map primes to their properties
    print(f"Mapping properties for {chunk_size} primes...")
    prime_data = map_prime_properties(primes, current_index)
    
    # Write this batch to the parquet file
    print(f"Writing batch to parquet file...")
    write_batch_to_parquet(parquet_writer, prime_data)
    
    # Update the current index for the next chunk
    current_index += chunk_size
    
    # Print progress
    elapsed = time.time() - start
    print(f"Progress: {total:_} primes processed in {elapsed:.2f}s ({total/elapsed:.2f} primes/s)")

# Close the parquet writer
parquet_writer.close()

print(f"Completed! {total:_} primes written to {output_file} in {time.time() - start:.2f}s")

File: generator/generator.go
package generator

import (
	"database/sql"
	"fmt"
	_ "github.com/marcboeker/go-duckdb"
	"github.com/xitongsys/parquet-go-source/local"
	"github.com/xitongsys/parquet-go/parquet"
	"github.com/xitongsys/parquet-go/writer"
	"math"
	"os"
	"runtime"
	"slices"
	"strconv"
	"sync"
	"time"
)

// GOAL: 1e9 in under 5 seconds
var GOAL int = 1e9
var upToSqrt int = int(math.Sqrt(float64(GOAL))) + 1
var CHUNK_SIZE int = GOAL / 100

func simpleSieve(limit int) []int {
	potPrimes := make([]bool, limit+1)
	var precomputedPrimes []int

	for i := 2; i <= limit; i++ {
		if potPrimes[i] == false {
			precomputedPrimes = append(precomputedPrimes, i)

			// Mark all multiples of i as non-prime
			for j := i * i; j <= limit; j += i {
				potPrimes[j] = true
			}
		}
	}

	return precomputedPrimes
}

func segmentedSieve(min int, max int, precomputedPrimes []int) []int {
	potPrimes := make([]bool, max-min+1)

	for _, precompPrime := range precomputedPrimes {
		firstMultiple := min / precompPrime

		if firstMultiple <= 1 {
			firstMultiple = precompPrime + precompPrime
		} else if (min % precompPrime) != 0 {
			firstMultiple = (firstMultiple * precompPrime) + precompPrime
		} else {
			firstMultiple = firstMultiple * precompPrime
		}

		for i := firstMultiple; i <= max; i += precompPrime {
			potPrimes[i-min] = true
		}
	}

	var finalPrimes []int
	for i := min; i <= max; i++ {
		if potPrimes[i-min] == false {
			finalPrimes = append(finalPrimes, i)
		}
	}

	return finalPrimes
}

func isPrime(prime int, primeList []int) bool {
	if prime <= 1 {
		return false
	}
	if prime == 2 || prime == 3 {
		return true
	}
	if prime%2 == 0 || prime%3 == 0 {
		return false
	}

	if prime <= upToSqrt {
		if slices.Contains(primeList, prime) {
			return true
		}
	}

	sqrtPrime := upToSqrt - 1
	for _, precompPrime := range primeList {
		if precompPrime > sqrtPrime {
			break // No need to check beyond sqrt(prime)
		}
		if prime%precompPrime == 0 {
			return false
		}
	}

	if len(primeList) > 0 && primeList[len(primeList)-1] < sqrtPrime {
		for i := primeList[len(primeList)-1] + 2; i <= sqrtPrime; i += 2 {
			if prime%i == 0 {
				return false
			}
		}
	}

	return true
}

func getNextTwinPrime(prime int, primeList []int) int {
	if isPrime(prime+2, primeList) {
		return prime + 2
	}

	return 0
}

func isSafePrime(prime int, primeList []int) bool {
	safeCandidate := (prime - 1) / 2

	if safeCandidate <= 1 {
		return false
	}

	if slices.Contains(primeList, safeCandidate) {
		return true
	}

	return isPrime(safeCandidate, primeList)
}

func getNForMersenne(prime int) int {
	log := math.Log2(float64(prime + 1))
	if log == math.Floor(log) {
		return int(log)
	} else {
		return 0
	}
}

func getDigitSum(prime, base int) int {
	if base < 2 {
		panic("base must be >= 2")
	}

	sum := 0
	for prime > 0 {
		sum += prime % base
		prime /= base
	}
	return sum
}

type PrimeData struct {
	Index          int32 `parquet:"name=index, type=INT32"`
	Prime          int64 `parquet:"name=prime, type=INT64"`
	GapToPrevious  int32 `parquet:"name=gap_to_previous, type=INT32"`
	TwinPrime      int32 `parquet:"name=twin_prime, type=INT32"`
	IsSafePrime    bool  `parquet:"name=is_safe_prime, type=BOOLEAN"`
	MersenneK      int32 `parquet:"name=mersenne_k, type=INT32"`
	DigitSumBase10 int32 `parquet:"name=digit_sum_base10, type=INT32"`
	DigitSumBase2  int32 `parquet:"name=digit_sum_base2, type=INT32"`
	DigitSumBase16 int32 `parquet:"name=digit_sum_base16, type=INT32"`
}

func CreateParquetWriter(outputFile string) (*writer.ParquetWriter, error) {
	pf, err := local.NewLocalFileWriter(outputFile)
	if err != nil {
		return nil, err
	}

	pw, err := writer.NewParquetWriter(pf, new(PrimeData), 4)
	if err != nil {
		return nil, err
	}

	pw.RowGroupSize = 128 * 1024 * 1024 // 128MB
	pw.CompressionType = parquet.CompressionCodec_SNAPPY

	return pw, nil
}

func WriteParquetBatch(pw *writer.ParquetWriter, data []PrimeData) error {
	for _, item := range data {
		if err := pw.Write(item); err != nil {
			return err
		}
	}

	return pw.Flush(true)
}

func CloseParquetWriter(pw *writer.ParquetWriter) error {
	return pw.WriteStop()
}

func GeneratePrimes() {
	start := time.Now()

	fmt.Printf("Starting data generation with a goal of %d...\n", GOAL)

	upToSqrt := int(math.Sqrt(float64(GOAL))) + 1

	fmt.Printf("Precomputing primes up to square root of goal: %d...\n", upToSqrt)

	precomputedPrimes := simpleSieve(upToSqrt)

	fmt.Println("Split goal into chunks of ranges...")

	chunks := getChunks(GOAL)

	fmt.Printf("%d chunks of ranges computed...\n", len(chunks))

	outputFile := "primes_data_unsorted.parquet"
	parquetWriter, err := CreateParquetWriter(outputFile)
	if err != nil {
		fmt.Printf("Error creating parquet writer: %v\n", err)
		return
	}
	defer CloseParquetWriter(parquetWriter)

	// Create a mutex for synchronizing writes to the parquet file
	var writerMutex sync.Mutex

	// Create a wait group to wait for all goroutines to finish
	var wg sync.WaitGroup

	maxConcurrent := runtime.NumCPU() // Adjust based on the number of CPU cores
	semaphore := make(chan struct{}, maxConcurrent)

	total := 0
	var totalMutex sync.Mutex

	// Process chunks in parallel
	for chunkIndex, chunk := range chunks {
		wg.Add(1)

		// Acquire semaphore
		semaphore <- struct{}{}

		// Launch a goroutine to process this chunk
		go func(index int, chunkRange [2]int) {
			defer wg.Done()
			defer func() { <-semaphore }() // Release semaphore when done

			min, max := chunkRange[0], chunkRange[1]
			fmt.Printf("Processing chunk %d/%d: [%d, %d]...\n", index+1, len(chunks), min, max)

			// Generate primes for this chunk
			primes := segmentedSieve(min, max, precomputedPrimes)
			chunkSize := len(primes)

			// Update total count (thread-safe)
			totalMutex.Lock()
			total += chunkSize
			totalMutex.Unlock()

			fmt.Printf("Mapping properties for %d primes in chunk %d...\n", chunkSize, index+1)
			primeData := mapPrimeProperties(primes, precomputedPrimes, 0) // Index doesn't matter for now

			fmt.Printf("Writing batch from chunk %d to parquet file...\n", index+1)
			writerMutex.Lock()
			err := WriteParquetBatch(parquetWriter, primeData)
			writerMutex.Unlock()

			if err != nil {
				fmt.Printf("Error writing batch from chunk %d: %v\n", index+1, err)
				return
			}

			elapsed := time.Since(start).Seconds()
			fmt.Printf("Chunk %d completed: %d primes processed in %.2fs\n", index+1, chunkSize, elapsed)
		}(chunkIndex, chunk)
	}

	// Wait for all goroutines to finish
	wg.Wait()

	elapsed := time.Since(start).Seconds()
	fmt.Printf("Completed! %d primes written to %s in %.2fs (%.2f primes/s)\n",
		total, outputFile, elapsed, float64(total)/elapsed)

	// Sort the final Parquet file using DuckDB
	fmt.Println("Sorting the final Parquet file using DuckDB...")
	sortParquetFile(outputFile, "primes_data.parquet")

	// Remove the unsorted file
	err = os.Remove(outputFile)
	if err != nil {
		fmt.Printf("Warning: Could not remove unsorted file: %v\n", err)
	} else {
		fmt.Println("Removed unsorted file.")
	}

	finalElapsed := time.Since(start).Seconds()
	fmt.Printf("All done! Total time: %.2fs\n", finalElapsed)
}

// sortParquetFile uses DuckDB Go library to sort a Parquet file by the prime column
func sortParquetFile(inputFile, outputFile string) error {
	// Open a new database connection (in-memory)
	db, err := sql.Open("duckdb", "")
	if err != nil {
		return fmt.Errorf("failed to open DuckDB connection: %v", err)
	}
	defer db.Close()

	// Create a query to read from the input Parquet file, sort by prime, and write to the output Parquet file
	query := fmt.Sprintf("COPY (SELECT * FROM '%s' ORDER BY prime) TO '%s' (FORMAT PARQUET);", inputFile, outputFile)

	// Execute the query
	_, err = db.Exec(query)
	if err != nil {
		return fmt.Errorf("failed to execute DuckDB query: %v", err)
	}

	fmt.Println("Parquet file sorted successfully.")
	return nil
}

func getChunks(limit int) [][2]int {
	var ranges [][2]int
	size := CHUNK_SIZE
	start := 2

	for start < limit {
		end := int(math.Min(float64(start+size), float64(limit)))
		ranges = append(ranges, [2]int{start, end})
		start += size
	}

	return ranges
}

func mapPrimeProperties(primes []int, precomputedPrimes []int, startIndex int) []PrimeData {
	// Pre-allocate the result slice
	data := make([]PrimeData, len(primes))

	// Convert startIndex to int32 once to avoid repeated conversions
	int32StartIndex := int32(startIndex)

	// Set the previous prime for gap calculation
	prevPrime := 2
	if startIndex > 0 && len(precomputedPrimes) > 0 {
		prevPrime = precomputedPrimes[len(precomputedPrimes)-1]
	}

	// Process in batches for better cache locality
	batchSize := 1024
	for batchStart := 0; batchStart < len(primes); batchStart += batchSize {
		batchEnd := batchStart + batchSize
		if batchEnd > len(primes) {
			batchEnd = len(primes)
		}

		// First pass: calculate indices, primes, and gaps (the expensive operations)
		for i := batchStart; i < batchEnd; i++ {
			prime := primes[i]
			pd := &data[i]
			pd.Index = int32StartIndex + int32(i)
			pd.Prime = int64(prime)
			pd.GapToPrevious = int32(prime - prevPrime)
			prevPrime = prime
		}

		// Second pass: calculate other properties
		for i := batchStart; i < batchEnd; i++ {
			prime := primes[i]
			pd := &data[i]

			// Twin prime calculation
			twinPrime := getNextTwinPrime(prime, precomputedPrimes)
			if twinPrime != 0 {
				pd.TwinPrime = int32(twinPrime)
			}

			// Safe prime calculation
			pd.IsSafePrime = isSafePrime(prime, precomputedPrimes)

			// Mersenne prime calculation
			n := getNForMersenne(prime)
			if (1<<n)-1 == prime {
				pd.MersenneK = int32(n)
			}

			// Digit sum calculations
			pd.DigitSumBase10 = int32(getDigitSumBase10(prime))
			pd.DigitSumBase2 = int32(getDigitSumBase2(prime))
			pd.DigitSumBase16 = int32(getDigitSumBase16(prime))
		}
	}

	return data
}

func getDigitSumBase10(n int) int {
	sum := 0
	for n > 0 {
		sum += n % 10
		n /= 10
	}
	return sum
}

func getDigitSumBase2(n int) int {
	return getDigitSum(n, 2)
}

func getDigitSumBase16(n int) int {
	sum := 0
	hexStr := strconv.FormatInt(int64(n), 16)

	for _, ch := range hexStr {
		if ch >= '0' && ch <= '9' {
			sum += int(ch - '0')
		} else if ch >= 'a' && ch <= 'f' {
			sum += int(ch - 'a' + 10)
		} else if ch >= 'A' && ch <= 'F' {
			sum += int(ch - 'A' + 10)
		}
	}

	return sum
}
